types/styled-components/index.d.ts

```diff
@@ -59,32 +60,47 @@ export type ThemedGlobalStyledClassProps<P, T> = P & {
 export interface GlobalStyleClass<P, T>
     extends React.ComponentClass<ThemedGlobalStyledClassProps<P, T>> {}
 
-export interface StyledComponentClass<P, T, O = P>
-    extends React.ComponentClass<ThemedOuterStyledProps<O, T>> {
+/** @deprecated Use StyledComponent instead */
+export type StyledComponentClass<P, T, O = {}> = StyledComponent<P, T, O>;
+
+// Make the StyledComponentInterpolation non-callable because typescript can't tell
+// between this and InterpolationFunction when inferring Interpolation
+type StyledComponentInterpolation = Pick<StyledComponent<any, any>, keyof StyledComponent<any, any>>;
+
+export interface StyledComponent<P, T, O = {}>
```

Comment:
To be honest, I don't really understand `O` at all. This was my attempt to fix a new `withComponent` test which was not forgetting the original component's props. I tried to reinterpret it as _only_ "additional props" instead of "props plus additional props", but I'm not sure I caught all the call sites. There might be some redundancy still around.